improc/simulator.py

```diff
@@ -0,0 +1,886 @@
+        Uses the optic_psf_mode to generate the PSF map
+        and calculate the width (FWHM).
+
+        Saves the results into optic_psf_image and optic_psf_fwhm.
+
+        """
+        self.oversampling = oversampling
+
+        if self.optic_psf_mode.lower() == 'gaussian':
+            self.optic_psf_image = make_gaussian(self.optic_psf_pars['sigma'] * self.oversampling)
+            self.optic_psf_fwhm = self.optic_psf_pars['sigma'] * 2.355
+        else:
+            raise ValueError(f'PSF mode not recognized: {self.optic_psf_mode}')
+
+    def make_vignette(self, imsize):
+        """
+        Make an image of the vignette part of the flat field.
+        Input the imsize as a tuple of (imsize_x, imsize_y).
+        """
+
+        v = np.ones((imsize[0], imsize[1]))
+        [xx, yy] = np.meshgrid(np.arange(-imsize[1] / 2, imsize[1] / 2), np.arange(-imsize[0] / 2, imsize[0] / 2))
+        xx += self.vignette_offset_x
+        yy += self.vignette_offset_y
+        rr = np.sqrt(xx ** 2 + yy ** 2)
+        v += (self.vignette_amplitude * (rr - self.vignette_radius)) ** 2
+        v[rr < self.vignette_radius] = 1.0
+        self.vignette_map = 1/v
+
+
+class SimSky:
+    """
+    Container for the properties of a simulated sky.
+    """
+    def __init__(self):
+        self.background_mean = None  # mean sky+dark current across image
+        self.background_std = None  # variation between images
+        self.background_instance = None  # the background for this specific image's sky
+
+        self.transmission_mean = None  # average sky transmission
+        self.transmission_std = None  # variation in transmission between images
+        self.transmission_instance = None  # the transmission for this specific image's sky
+
+        self.seeing_mean = None  # the average seeing in this survey
+        self.seeing_std = None  # the variation in seeing between images
+        self.seeing_instance = None  # the seeing for this specific image's sky
+
+        self.oversampling = None  # how much do we need the PSF to be oversampled?
+        self.atmos_psf_mode = None  # e.g., 'gaussian'
+        self.atmos_psf_pars = None  # a dict with the parameters used to make this PSF
+        self.atmos_psf_image = None  # the final shape of the PSF for this image
+        self.atmos_psf_fwhm = None  # e.g., the seeing
+
+    def make_atmos_psf(self, oversampling):
+        """
+        Use the psf mode, pars and the seeing_instance
+        to produce an atmospheric PSF.
+        This PSF is used to convolve the optical PSF to get the total PSF.
+
+        """
+        self.oversampling = oversampling
+
+        if self.atmos_psf_mode.lower() == 'gaussian':
+            self.atmos_psf_image = make_gaussian(self.atmos_psf_pars['sigma'] * self.oversampling)
+            self.atmos_psf_fwhm = self.atmos_psf_pars['sigma'] * 2.355
+        else:
+            raise ValueError(f'PSF mode not recognized: {self.atmos_psf_mode}')
+
+
+class SimStars:
```

Comment:
Obviously not for this PR, but: do you have any ambitions for SimGalaxies?  That's a pretty gigantic project, and it's probably worth looking into GalSim or something like that which has already done simulations of galaxy profiles.

Reply:
I think that might be a nice addition (as an optional parameter for the simulator). But really you can get a lot of the way there just by putting some elongated Gaussians as stand-ins for galaxies. 
The same is probably true for cosmic rays and bleeds. In terms of figuring out sky flat algorithms (and maybe other things) it may be close enough that it wouldn't matter if we hack something really simple together (like the very simple Gaussian PSF I am using).